File: learn/tree/avl-tree/avl_tree.go


type AVLNode struct {
	data   int
	height int
	left   *AVLNode
	right  *AVLNode
}

type AVLTree struct {
	root *AVLNode
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func (t *AVLTree) height(N *AVLNode) int {
	if N == nil {
		return 0
	}
	return N.height
}

func (t *AVLTree) rightRotate(y *AVLNode) *AVLNode {
	x := y.left
	T2 := x.right
	x.right = y
	y.left = T2
	y.height = max(t.height(y.left), t.height(y.right)) + 1
	x.height = max(t.height(x.left), t.height(x.right)) + 1
	return x
}

func (t *AVLTree) leftRotate(x *AVLNode) *AVLNode {
	y := x.right
	T2 := y.left
	y.left = x
	x.right = T2
	x.height = max(t.height(x.left), t.height(x.right)) + 1
	y.height = max(t.height(y.left), t.height(y.right)) + 1
	return y
}

func (t *AVLTree) getBalance(N *AVLNode) int {
	if N == nil {
		return 0
	}
	return t.height(N.left) - t.height(N.right)
}

func (t *AVLTree) Insert(data int) {
	t.root = t.insertNode(t.root, data)
}

func (t *AVLTree) insertNode(node *AVLNode, data int) *AVLNode {
	if node == nil {
		return &AVLNode{data: data, height: 1}
	}
	if data < node.data {
		node.left = t.insertNode(node.left, data)
	} else if data > node.data {
		node.right = t.insertNode(node.right, data)
	} else {
		return node
	}
	node.height = 1 + max(t.height(node.left), t.height(node.right))
	balance := t.getBalance(node)
	if balance > 1 && data < node.left.data {
		return t.rightRotate(node)
	}
	if balance < -1 && data > node.right.data {
		return t.leftRotate(node)
	}
	if balance > 1 && data > node.left.data {
		node.left = t.leftRotate(node.left)
		return t.rightRotate(node)
	}
	if balance < -1 && data < node.right.data {
		node.right = t.rightRotate(node.right)
		return t.leftRotate(node)
	}
	return node
}

File: learn/tree/avl-tree/avl_tree.rs
use std::cmp;

pub struct AVLNode {
    pub data: i32,
    pub height: i32,
    pub left: Option<Box<AVLNode>>,
    pub right: Option<Box<AVLNode>>,
}

pub struct AVLTree {
    pub root: Option<Box<AVLNode>>,
}

impl AVLTree {
    pub fn new() -> Self {
        AVLTree { root: None }
    }

    fn height(node: &Option<Box<AVLNode>>) -> i32 {
        node.as_ref().map_or(0, |n| n.height)
    }

    fn update_height(node: &mut AVLNode) {
        node.height = 1 + cmp::max(Self::height(&node.left), Self::height(&node.right));
    }

    fn rotate_right(mut y: Box<AVLNode>) -> Box<AVLNode> {
        let mut x = y.left.take().unwrap();
        y.left = x.right.take();
        Self::update_height(&mut y);
        x.right = Some(y);
        Self::update_height(&mut x);
        x
    }

    fn rotate_left(mut x: Box<AVLNode>) -> Box<AVLNode> {
        let mut y = x.right.take().unwrap();
        x.right = y.left.take();
        Self::update_height(&mut x);
        y.left = Some(x);
        Self::update_height(&mut y);
        y
    }

    pub fn insert(&mut self, data: i32) {
        self.root = Self::insert_node(self.root.take(), data);
    }

    fn insert_node(node: Option<Box<AVLNode>>, data: i32) -> Option<Box<AVLNode>> {
        let mut node = match node {
            None => {
                return Some(Box::new(AVLNode {
                    data,
                    height: 1,
                    left: None,
                    right: None,
                }))
            }
            Some(n) => n,
        };

        if data < node.data {
            node.left = Self::insert_node(node.left.take(), data);
        } else if data > node.data {
            node.right = Self::insert_node(node.right.take(), data);
        } else {
            return Some(node);
        }

        Self::update_height(&mut node);
        let balance = Self::height(&node.left) - Self::height(&node.right);

        if balance > 1 {
            let left_bal = node
                .left
                .as_ref()
                .map_or(0, |l| Self::height(&l.left) - Self::height(&l.right));
            if left_bal < 0 {
                node.left = Some(Self::rotate_left(node.left.take().unwrap()));
            }
            return Some(Self::rotate_right(node));
        }

        if balance < -1 {
            let right_bal = node
                .right
                .as_ref()
                .map_or(0, |r| Self::height(&r.left) - Self::height(&r.right));
            if right_bal > 0 {
                node.right = Some(Self::rotate_right(node.right.take().unwrap()));
            }
            return Some(Self::rotate_left(node));
        }

        Some(node)
    }
}
